gr-iio/lib/device_source_impl.cc

```diff
@@ -0,0 +1,409 @@
+                                        const std::string& device_phy,
+                                        const std::vector<std::string>& params,
+                                        unsigned int buffer_size,
+                                        unsigned int decimation)
+{
+    return gnuradio::get_initial_sptr(
+        new device_source_impl(device_source_impl::get_context(uri),
+                               true,
+                               device,
+                               channels,
+                               device_phy,
+                               params,
+                               buffer_size,
+                               decimation));
+}
+
+device_source::sptr device_source::make_from(struct iio_context* ctx,
+                                             const std::string& device,
+                                             const std::vector<std::string>& channels,
+                                             const std::string& device_phy,
+                                             const std::vector<std::string>& params,
+                                             unsigned int buffer_size,
+                                             unsigned int decimation)
+{
+    return gnuradio::get_initial_sptr(new device_source_impl(
+        ctx, false, device, channels, device_phy, params, buffer_size, decimation));
+}
+
+void device_source_impl::set_params(struct iio_device* phy,
+                                    const std::vector<std::string>& params)
+{
+    for (std::vector<std::string>::const_iterator it = params.begin(); it != params.end();
+         ++it) {
+        struct iio_channel* chn = NULL;
+        const char* attr = NULL;
+        size_t pos;
+        int ret;
+
+        pos = it->find('=');
+        if (pos == std::string::npos) {
+            std::cerr << "Misformed line: " << *it << std::endl;
+            continue;
+        }
+
+        std::string key = it->substr(0, pos);
+        std::string val = it->substr(pos + 1, std::string::npos);
+
+        ret = iio_device_identify_filename(phy, key.c_str(), &chn, &attr);
+        if (ret) {
+            std::cerr << "Parameter not recognized: " << key << std::endl;
+            continue;
+        }
+
+        if (chn)
+            ret = iio_channel_attr_write(chn, attr, val.c_str());
+        else if (iio_device_find_attr(phy, attr))
+            ret = iio_device_attr_write(phy, attr, val.c_str());
+        else
+            ret = iio_device_debug_attr_write(phy, attr, val.c_str());
+        if (ret < 0) {
+            std::cerr << "Unable to write attribute " << key << ": " << ret << std::endl;
+        }
+    }
+}
+
+void device_source_impl::set_params(const std::vector<std::string>& params)
+{
+    set_params(this->phy, params);
+}
+
+void device_source_impl::set_buffer_size(unsigned int _buffer_size)
+{
+    boost::unique_lock<boost::mutex> lock(iio_mutex);
```

Comment:
This is not a showstopper for merging IMO, but there's a lot of unnecessary boost usage here.